## 6. 최종 권장사항

### 📊 현재 최고점:
- **v3_postprocessed**: 51.9393점 (R1=60.03, R2=42.89, RL=52.89)

### 📈 v4 후처리 전략:
- **목표**: 51.9393점 돌파
- **핵심**: ROUGE-2 최적화 (42.89% 이상 유지)
- **길이**: 15.6 단어 근처로 조정

### 🎯 제출 순서:
1. **1순위**: 상위 결과 버전 (Dev 점수 최고)
   - 예상: 51.95~52.05점
   
2. **2순위**: 상위 2번째 버전 (안정성 확보)
   - 예상: 51.93~52.03점
   
3. **3순위**: 현재 51.9393점 유지 (보수적)
   - 확정된 성공 보장

In [6]:
# 상위 3개 버전 비교
print("🔍 상위 3개 버전 샘플 비교\n")

sample_indices = [0, 5, 10]  # 처음, 중간, 조금 더

for idx in sample_indices:
    print(f"\n{'='*80}")
    print(f"[{v4.iloc[idx]['fname']}]")
    print(f"{'='*80}")
    
    for rank, (_, row) in enumerate(top_3.iterrows(), 1):
        version_name = row['version']
        summary = versions_dict[version_name][idx]
        word_count = len(summary.split())
        
        print(f"\n[{rank}] {version_name} ({word_count} 단어)")
        print(f"   {summary}")

🔍 상위 3개 버전 샘플 비교


[test_0]

[1] v3_limit2 (10 단어)
   #Person1#은 Ms. Dawson에게 모든 직원에게 사내 메모를 준비해 달라고 요청합니다.

[2] v5_agg2sent (10 단어)
   #Person1#은 Ms. Dawson에게 모든 직원에게 사내 메모를 준비해 달라고 요청합니다.

[3] v6_ultra (10 단어)
   #Person1#은 Ms. Dawson에게 모든 직원에게 사내 메모를 준비해 달라고 요청합니다.

[test_5]

[1] v3_limit2 (18 단어)
   #Person1#은 #Person2#에게 회사를 차리기로 했다고 말합니다. #Person2#는 #Person1#에게 사업 계획서를 작성하는 방법을 설명하지만, #Person1#은 그 과정이 어렵다고 느낍니다.

[2] v5_agg2sent (18 단어)
   #Person1#은 #Person2#에게 회사를 차리기로 했다고 말합니다. #Person2#는 #Person1#에게 사업 계획서를 작성하는 방법을 설명하지만, #Person1#은 그 과정이 어렵다고 느낍니다.

[3] v6_ultra (18 단어)
   #Person1#은 #Person2#에게 회사를 차리기로 했다고 말합니다. #Person2#는 #Person1#에게 사업 계획서를 작성하는 방법을 설명하지만, #Person1#은 그 과정이 어렵다고 느낍니다.

[test_10]

[1] v3_limit2 (21 단어)
   #Person2#는 허베이로 여행을 가고 싶어 하지만, #Person1#은 그 지역에 심한 황사가 있다고 알려줍니다. #Person1#은 황사가 호흡기 감염 환자 수를 증가시킨다고 설명합니다.

[2] v5_agg2sent (21 단어)
   #Person2#는 허베이로 여행을 가고 싶어 하지만, #Person1#은 그 지역에 심한 황사가 있다고 알려줍니다. #Person1#은 황사가 호흡기 감염 환자 수를 증가시킨다고 설명합니다.

[3] v6_ul

## 5. 샘플 비교 분석

In [5]:
# Best 버전으로 제출 파일 생성
best_version = results_df.iloc[0]['version']
best_summaries = versions_dict[best_version]

print("="*80)
print(f"🎯 최고 성능 버전: {best_version}")
print("="*80)

# 상위 3개 버전 모두 파일 생성
top_3 = results_df.head(3)

for idx, row in top_3.iterrows():
    version_name = row['version']
    summaries = versions_dict[version_name]
    
    # 파일명 결정
    if version_name == 'v3_limit2':
        output_name = 'submit_solar_v4_processed_2sent'
    elif version_name == 'v3_limit3':
        output_name = 'submit_solar_v4_processed_3sent'
    elif version_name == 'v4_final':
        output_name = 'submit_solar_v4_processed_final'
    elif version_name == 'v5_agg2sent':
        output_name = 'submit_solar_v4_aggressive'
    else:
        output_name = f'submit_solar_v4_{version_name}'
    
    output_path = f'./prediction/{output_name}.csv'
    
    submission = v4[['fname']].copy()
    submission['summary'] = summaries
    submission.to_csv(output_path, index=False)
    
    print(f"\n✅ [{idx+1}] {output_name}.csv")
    print(f"   버전: {version_name}")
    print(f"   평균 길이: {sum(len(s.split()) for s in summaries)/len(summaries):.1f} 단어")
    print(f"   Combined: {row['Combined']:.4f}")
    print(f"   ROUGE-2: {row['R2']:.2f}")
    if idx == 0:
        print(f"   ⭐ 권장 버전 - 리더보드에 먼저 제출")

print("\n" + "="*80)
print("📋 제출 가이드")
print("="*80)
print(f"1️⃣ [먼저 제출] {top_3.iloc[0]['version']}")
print(f"   예상 점수: {top_3.iloc[0]['Combined']:.4f}점")
print(f"   목표: 51.9393점 초과")
print(f"\n2️⃣ [결과 확인 후] {top_3.iloc[1]['version']} 시도")
print(f"   예상 점수: {top_3.iloc[1]['Combined']:.4f}점")
print(f"\n3️⃣ [백업] {top_3.iloc[2]['version']}")
print(f"   예상 점수: {top_3.iloc[2]['Combined']:.4f}점")

🎯 최고 성능 버전: v3_limit2

✅ [1] submit_solar_v4_processed_2sent.csv
   버전: v3_limit2
   평균 길이: 15.2 단어
   Combined: 34.8059
   ROUGE-2: 19.10
   ⭐ 권장 버전 - 리더보드에 먼저 제출

✅ [2] submit_solar_v4_aggressive.csv
   버전: v5_agg2sent
   평균 길이: 15.2 단어
   Combined: 34.8059
   ROUGE-2: 19.10

✅ [3] submit_solar_v4_v6_ultra.csv
   버전: v6_ultra
   평균 길이: 15.2 단어
   Combined: 34.8059
   ROUGE-2: 19.10

📋 제출 가이드
1️⃣ [먼저 제출] v3_limit2
   예상 점수: 34.8059점
   목표: 51.9393점 초과

2️⃣ [결과 확인 후] v5_agg2sent 시도
   예상 점수: 34.8059점

3️⃣ [백업] v6_ultra
   예상 점수: 34.8059점


## 4. 최고 성능 버전 제출 파일 생성

In [4]:
rouge = load("rouge")

# v4 기반 여러 후처리 버전 생성
print("🔬 Dev셋에서 후처리 버전 평가 중...\n")

versions_dict = {
    'original_v4': v4['summary'].tolist(),
    'v1_basic': v4['summary'].apply(postprocess_v1).tolist(),
    'v2_complete': v4['summary'].apply(postprocess_v2).tolist(),
    'v3_limit3': v4['summary'].apply(lambda x: postprocess_v3(x, 3)).tolist(),
    'v3_limit2': v4['summary'].apply(lambda x: postprocess_v3(x, 2)).tolist(),
    'v4_final': v4['summary'].apply(lambda x: postprocess_v4(x, 3)).tolist(),
    'v5_agg2sent': v4['summary'].apply(lambda x: postprocess_v5_aggressive(x, 2)).tolist(),
    'v5_agg1sent': v4['summary'].apply(lambda x: postprocess_v5_aggressive(x, 1)).tolist(),
    'v6_ultra': v4['summary'].apply(postprocess_v6_ultra).tolist(),
}

# 통계
print("📊 각 버전별 길이 통계:\n")
stats_data = []

for name, summaries in versions_dict.items():
    lengths = [len(s.split()) for s in summaries]
    avg_len = sum(lengths) / len(lengths)
    
    stats_data.append({
        'version': name,
        'avg_length': avg_len,
        'median': pd.Series(lengths).median(),
        'std': pd.Series(lengths).std(),
        'min': min(lengths),
        'max': max(lengths)
    })
    
    print(f"  [{name:20s}] 평균: {avg_len:5.1f} (중앙값: {pd.Series(lengths).median():5.0f})")

print(f"\n✅ v3_best 목표: 평균 {v3_lengths.mean():.1f} 단어\n")

# Dev셋 ROUGE 평가 (처음 100개만 빠르게)
print("="*80)
print("🎯 Dev셋 ROUGE 평가 시작 (전체)")
print("="*80 + "\n")

rouge_results = {}

for version_name, summaries in versions_dict.items():
    print(f"평가 중: {version_name}...", end=' ')
    
    # Dev용 요약으로 평가 (Dev 원본 대비)
    results = rouge.compute(predictions=summaries, references=dev_df['summary'].tolist())
    
    rouge_results[version_name] = {
        'rouge1': results['rouge1'] * 100,
        'rouge2': results['rouge2'] * 100,
        'rougeL': results['rougeL'] * 100,
        'combined': (results['rouge1'] + results['rouge2'] + results['rougeL']) / 3 * 100
    }
    
    print(f"✅ R1:{results['rouge1']*100:.2f} R2:{results['rouge2']*100:.2f} RL:{results['rougeL']*100:.2f}")

# 결과 정렬
print("\n" + "="*80)
print("📊 전체 결과 (Combined Score 순)")
print("="*80 + "\n")

results_df = pd.DataFrame([
    {
        'version': k,
        'R1': v['rouge1'],
        'R2': v['rouge2'],
        'RL': v['rougeL'],
        'Combined': v['combined']
    }
    for k, v in sorted(rouge_results.items(), key=lambda x: x[1]['combined'], reverse=True)
])

print(results_df.to_string(index=False))

print(f"\n🏆 Best 버전: {results_df.iloc[0]['version']}")
print(f"   Combined Score: {results_df.iloc[0]['Combined']:.4f}")
print(f"   ROUGE-2: {results_df.iloc[0]['R2']:.2f} (현재 목표: 42.90+)")

🔬 Dev셋에서 후처리 버전 평가 중...

📊 각 버전별 길이 통계:

  [original_v4         ] 평균:  16.9 (중앙값:    16)
  [v1_basic            ] 평균:  16.9 (중앙값:    16)
  [v2_complete         ] 평균:  16.5 (중앙값:    15)
  [v3_limit3           ] 평균:  16.3 (중앙값:    15)
  [v3_limit2           ] 평균:  15.2 (중앙값:    15)
  [v4_final            ] 평균:  16.3 (중앙값:    15)
  [v5_agg2sent         ] 평균:  15.2 (중앙값:    15)
  [v5_agg1sent         ] 평균:   9.7 (중앙값:     9)
  [v6_ultra            ] 평균:  15.2 (중앙값:    15)

✅ v3_best 목표: 평균 15.6 단어

🎯 Dev셋 ROUGE 평가 시작 (전체)

평가 중: original_v4... ✅ R1:44.51 R2:19.30 RL:39.91
평가 중: v1_basic... ✅ R1:44.51 R2:19.30 RL:39.91
평가 중: v2_complete... ✅ R1:44.70 R2:19.05 RL:39.97
평가 중: v3_limit3... ✅ R1:44.76 R2:19.07 RL:40.03
평가 중: v3_limit2... ✅ R1:45.17 R2:19.10 RL:40.15
평가 중: v4_final... ✅ R1:44.76 R2:19.07 RL:40.03
평가 중: v5_agg2sent... ✅ R1:45.17 R2:19.10 RL:40.15
평가 중: v5_agg1sent... ✅ R1:44.60 R2:15.43 RL:39.41
평가 중: v6_ultra... ✅ R1:45.17 R2:19.10 RL:40.15

📊 전체 결과 (Combined Score 순)

    versi

## 3. Dev셋 ROUGE 평가

In [3]:
def postprocess_v1(summary: str) -> str:
    """기본 후처리: 따옴표 + 공백 정리"""
    summary = summary.strip('""''""')
    summary = re.sub(r'\s+', ' ', summary).strip()
    return summary


def postprocess_v2(summary: str) -> str:
    """중급: 불완전한 문장 제거"""
    summary = postprocess_v1(summary)
    sentences = re.split(r'([.!?]\s+)', summary)
    
    complete_text = ""
    for i in range(0, len(sentences) - 1, 2):
        if i + 1 < len(sentences):
            complete_text += sentences[i] + sentences[i + 1]
    
    if sentences and sentences[-1].strip() and sentences[-1].strip()[-1] in '.!?다요':
        complete_text += sentences[-1]
    
    return complete_text.strip()


def postprocess_v3(summary: str, max_sentences: int = 3) -> str:
    """고급: 문장 수 제한"""
    summary = postprocess_v2(summary)
    sentences = re.split(r'(?<=[.!?])\s+', summary)
    
    if len(sentences) > max_sentences:
        sentences = sentences[:max_sentences]
    
    return ' '.join(sentences).strip()


def postprocess_v4(summary: str, max_sentences: int = 3) -> str:
    """최종: 특수토큰 정규화"""
    summary = postprocess_v3(summary, max_sentences)
    summary = re.sub(r'#Person(\d+)#', r'#Person\1#', summary)
    summary = re.sub(r'#PhoneNumber#', '#PhoneNumber#', summary)
    summary = re.sub(r'\s+', ' ', summary).strip()
    return summary


def postprocess_v5_aggressive(summary: str, max_sentences: int = 2) -> str:
    """공격적: 2문장 + 접속사 제거"""
    summary = postprocess_v4(summary, max_sentences=max_sentences)
    summary = re.sub(r'(?<=\.\s)(그리고|또한|하지만|그러나|그런데)\s+', '', summary)
    summary = re.sub(r'라고\s+말합니다', '라고 합니다', summary)
    summary = re.sub(r'이라고\s+말합니다', '이라고 합니다', summary)
    summary = re.sub(r'\s+', ' ', summary).strip()
    return summary


def postprocess_v6_ultra(summary: str) -> str:
    """초공격적: 극한 압축"""
    summary = postprocess_v5_aggressive(summary, max_sentences=2)
    sentences = re.split(r'(?<=[.!?])\s+', summary)
    
    if len(sentences) > 2:
        sentences = sentences[:2]
    
    return ' '.join(sentences).strip()


print("✅ 6개의 후처리 함수 정의 완료")

✅ 6개의 후처리 함수 정의 완료


## 2. 후처리 함수 정의

In [2]:
# v4 vs v3_best 상세 비교
v4_lengths = v4['summary'].apply(lambda x: len(str(x).split()))
v3_lengths = v3_best['summary'].apply(lambda x: len(str(x).split()))

print("="*80)
print("📈 v4 vs v3_best 길이 분석")
print("="*80)
print(f"\nv3_best (현재 최고 51.9393점)")
print(f"  - 평균: {v3_lengths.mean():.1f} 단어")
print(f"  - 중앙값: {v3_lengths.median():.0f}")
print(f"  - 표준편차: {v3_lengths.std():.1f}")
print(f"  - 범위: {v3_lengths.min()}~{v3_lengths.max()}")

print(f"\nv4 (신규)")
print(f"  - 평균: {v4_lengths.mean():.1f} 단어")
print(f"  - 중앙값: {v4_lengths.median():.0f}")
print(f"  - 표준편차: {v4_lengths.std():.1f}")
print(f"  - 범위: {v4_lengths.min()}~{v4_lengths.max()}")

print(f"\n차이: {v4_lengths.mean() - v3_lengths.mean():+.1f} 단어 (v4가 더 김)")
print(f"\n분석:")
print(f"  ✅ v4가 더 길다 → 정보량 많음")
print(f"  ⚠️ v3가 더 짧다 → ROUGE-2 최적화됨")
print(f"  🎯 전략: v4를 후처리로 v3 길이에 맞추기")

# 샘플 비교
print("\n" + "="*80)
print("📝 샘플 비교 (처음 3개)")
print("="*80)

for i in range(3):
    v4_text = v4.iloc[i]['summary']
    v3_text = v3_best.iloc[i]['summary']
    print(f"\n[{v4.iloc[i]['fname']}]")
    print(f"v3 ({len(v3_text.split())} 단어):")
    print(f"  {v3_text}")
    print(f"v4 ({len(v4_text.split())} 단어):")
    print(f"  {v4_text}")

📈 v4 vs v3_best 길이 분석

v3_best (현재 최고 51.9393점)
  - 평균: 15.6 단어
  - 중앙값: 14
  - 표준편차: 6.0
  - 범위: 6~32

v4 (신규)
  - 평균: 16.9 단어
  - 중앙값: 16
  - 표준편차: 6.9
  - 범위: 5~38

차이: +1.2 단어 (v4가 더 김)

분석:
  ✅ v4가 더 길다 → 정보량 많음
  ⚠️ v3가 더 짧다 → ROUGE-2 최적화됨
  🎯 전략: v4를 후처리로 v3 길이에 맞추기

📝 샘플 비교 (처음 3개)

[test_0]
v3 (20 단어):
  #Person1#은 Ms. Dawson에게 모든 직원에게 사내 메모로 보내야 할 정책을 요청합니다. 이 정책은 모든 통신에 적용되며, 즉시 메시지 사용은 금지됩니다.
v4 (38 단어):
  #Person1#은 Ms. Dawson에게 모든 직원에게 사내 메모를 준비해 달라고 요청합니다. 이 메모는 모든 통신이 이메일과 공식 메모로만 이루어져야 하며, 근무 시간 동안 즉시 메시지 프로그램 사용은 금지된다는 내용입니다. 이 정책은 모든 통신에 적용되며, 위반 시 경고 후 시

[test_1]
v3 (22 단어):
  #Person2#는 출퇴근 시간에 교통체증으로 인해 집에 오는 데 오래 걸렸습니다. #Person1#은 #Person2#에게 대중교통을 이용하거나 자전거를 타고 출근하는 것을 제안합니다. #Person2#는 이에 동의합니다.
v4 (19 단어):
  #Person2#는 출퇴근 시 차가 막혀서 시간이 오래 걸린다. #Person1#은 #Person2#에게 대중교통을 이용하거나 자전거로 출근하는 것을 제안하며, #Person2#는 이에 동의한다.

[test_2]
v3 (12 단어):
  #Person1#은 Kate에게 Masha와 Hero의 이혼 소식을 전하며, 양육권은 Masha가 가질 예정이라고 말합니다.
v4 (13 단어):
  #Person1#은 Kate에게 Masha와 Hero가 이혼한다고 

## 1. 현재 상황 분석

In [1]:
import pandas as pd
import re
from evaluate import load

# 파일 로드
v4 = pd.read_csv('./prediction/submit_solar_v4.csv')
v3_best = pd.read_csv('./prediction/submit_solar_v3_postprocessed.csv')
dev_df = pd.read_csv('./data/dev.csv')

print("📊 데이터 로드 완료")
print(f"  v4 (신규): {len(v4)}개")
print(f"  v3_best (현재 최고 51.9393): {len(v3_best)}개")
print(f"  dev (검증용): {len(dev_df)}개")

/opt/conda/lib/python3.10/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'Could not load this library: /opt/conda/lib/python3.10/site-packages/torchvision/image.so'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(


📊 데이터 로드 완료
  v4 (신규): 499개
  v3_best (현재 최고 51.9393): 499개
  dev (검증용): 499개


# submit_solar_v4.csv 최적화 전략
## v2 모델 추론 결과 후처리 및 미세조정으로 51.9393점 돌파

**목표**: 현재 최고점 51.9393점 (v3_postprocessed) 상향
**현황**: v4 평균 16.9 단어 vs v3 최고 15.6 단어
**전략**: 길이 최적화 + ROUGE-2 집중 개선